### ISSUE TITLE
Problems with android system back not dismissing the sheet

### ISSUE THREAD

**Issue by StylianosGakis** [oleksandrbalan/modalsheet#36]
When using the sheet normally, I've had problems with the `onSystemBack = { // this }` not being called at all when the system back is pressed, and instead it is simply doing what it'd do if the sheet wasn't there at all.

I am asking here just to see if this is some known issue, if not I probably should try to make a repro to better explain what I am experiencing and you might just be able to help me by pointing out any obvious mistakes I may be making.

**Comment by oleksandrbalan** [comment]
Huh, that's strange 🤔 

Do you see the same behavior in demo app?

**Comment by StylianosGakis** [comment]
Hmm turns out I really couldn't repro this in a smaller project.
It may have something to do with the fact that we are also inside a NavHost where we're using the sheet or something like that. And it's good to hear that you are not aware of any such problems from other people using it.

I think the right thing for me to do is to just close this, and if I manage to make a minimal repro I will come back here. If it turns out to have been just something we do wrong somehow then sorry for bothering 😊

**Comment by StylianosGakis** [comment]
Whoops, immediately as I wrote this I remembered one last thing I did not test before.
Having `android:enableOnBackInvokedCallback="true"` in the manifest to support predictive back.
Adding that, then having a simple app like:
```
class MainActivity : ComponentActivity() {
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    enableEdgeToEdge()
    setContent {
      BottomSheetBackCheckTheme {
        Surface(Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background) {
          App()
        }
      }
    }
  }
}

@OptIn(ExperimentalSheetApi::class)
@Composable
fun App(modifier: Modifier = Modifier) {
  var visible by remember { mutableStateOf(false) }
  ModalSheet(
    visible = visible,
    onVisibleChange = { visible = it },
    onSystemBack = { visible = false },
  ) {
    Text("Sheet content")
  }
  BackHandler(true) {
    println("Oh no")
  }
  Box(modifier, Alignment.Center) {
    Button({ visible = true }) {
      Text("Open")
    }
  }
}
```
Repros this 100% of the time.

For context, this is running on API 35. But I see the same on API 33 too, since I think that it is when predictive back was introduced? This probably then is a problem for all Android versions which have predictive back enabled.

The behavior btw is that `onSystemBack` just is not called at all.

**Comment by StylianosGakis** [comment]
Looking at the docs here https://developer.android.com/guide/navigation/custom-back/predictive-back-gesture there is an interesting note which says:
"Note: KeyEvent.KEYCODE_BACK is not deprecated as there are some supported use cases of KeyEvent.KEYCODE_BACK; however, intercepting back events from KeyEvent.KEYCODE_BACK is no longer supported."
So there needs to be changes internally in the library code to support this use case it seems

**Comment by StylianosGakis** [comment]
Applying this patch (aka just adding enableOnBackInvokedCallback) 
```
Subject: [PATCH] Bug repro
---
Index: demo/src/main/AndroidManifest.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/demo/src/main/AndroidManifest.xml b/demo/src/main/AndroidManifest.xml
--- a/demo/src/main/AndroidManifest.xml	(revision 55d0c458f612dc2fb1c601d7b45333aee2c97900)
+++ b/demo/src/main/AndroidManifest.xml	(date 1719927983906)
@@ -3,6 +3,7 @@
 
     <application
         android:allowBackup="true"
+        android:enableOnBackInvokedCallback="true"
         android:icon="@mipmap/ic_launcher"
         android:label="@string/app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
```
To the demo app inside this repository and running the app results in this behavior:

https://github.com/oleksandrbalan/modalsheet/assets/44558292/b715a68d-9b72-4599-9478-67a2b0735b8b

The back gesture is forwarded up to the activity above (in this case your BackHandler to close the specific demo screen), the sheet does not see it at all. The only reason it is dismissed is that after the CrossFade finishes, the ModalSheet leaves composition completely, so it dismisses itself immediately (without an animation) due to just not being in composition anymore.

**Comment by StylianosGakis** [comment]
Just as a POC, this patch
```
Index: modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt b/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt
--- a/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt	(revision 9df1393bfc0517554632f3d247f467e24f5de2dd)
+++ b/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt	(date 1719929199943)
@@ -4,9 +4,12 @@
 import android.app.Activity
 import android.content.Context
 import android.content.ContextWrapper
+import android.os.Build
 import android.view.KeyEvent
 import android.view.View
 import android.view.ViewGroup
+import android.window.OnBackInvokedCallback
+import android.window.OnBackInvokedDispatcher
 import androidx.compose.foundation.layout.Box
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.CompositionContext
@@ -51,30 +54,31 @@
 @Composable
 internal fun FullscreenPopup(
     onSystemBack: (() -> Unit)? = null,
-    content: @Composable () -> Unit
+    content: @Composable () -> Unit,
 ) {
     val view = LocalView.current
     val parentComposition = rememberCompositionContext()
     val currentContent by rememberUpdatedState(content)
     val popupId = rememberSaveable { UUID.randomUUID() }
-    val popupLayout = remember {
-        PopupLayout(
-            onSystemBack = onSystemBack,
-            composeView = view,
-            popupId = popupId
-        ).apply {
-            setContent(parentComposition) {
-                Box(Modifier.semantics { this.popup() }) {
-                    currentContent()
-                }
-            }
-        }
-    }
+    val popupLayout =
+        remember {
+            PopupLayout(
+                onSystemBack = onSystemBack,
+                composeView = view,
+                popupId = popupId,
+            ).apply {
+                setContent(parentComposition) {
+                    Box(Modifier.semantics { this.popup() }) {
+                        currentContent()
+                    }
+                }
+            }
+        }
 
     DisposableEffect(popupLayout) {
         popupLayout.show()
         popupLayout.updateParameters(
-            onSystemBack = onSystemBack
+            onSystemBack = onSystemBack,
         )
         onDispose {
             popupLayout.disposeComposition()
@@ -85,7 +89,7 @@
 
     SideEffect {
         popupLayout.updateParameters(
-            onSystemBack = onSystemBack
+            onSystemBack = onSystemBack,
         )
     }
 }
@@ -97,10 +101,9 @@
 private class PopupLayout(
     private var onSystemBack: (() -> Unit)?,
     composeView: View,
-    popupId: UUID
+    popupId: UUID,
 ) : AbstractComposeView(composeView.context),
     ViewRootForInspector {
-
     private val decorView = findOwner<Activity>(composeView.context)?.window?.decorView as ViewGroup
 
     override val subCompositionView: AbstractComposeView get() = this
@@ -127,13 +130,16 @@
         decorView.addView(
             this,
             0,
-            MarginLayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
+            MarginLayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT),
         )
 
         requestFocus()
     }
 
-    fun setContent(parent: CompositionContext, content: @Composable () -> Unit) {
+    fun setContent(
+        parent: CompositionContext,
+        content: @Composable () -> Unit,
+    ) {
         setParentCompositionContext(parent)
         this.content = content
         shouldCreateCompositionOnAttachedToWindow = true
@@ -146,34 +152,49 @@
 
     @Suppress("ReturnCount")
     override fun dispatchKeyEvent(event: KeyEvent): Boolean {
-        if (event.keyCode == KeyEvent.KEYCODE_BACK && onSystemBack != null) {
-            if (keyDispatcherState == null) {
-                return super.dispatchKeyEvent(event)
-            }
-            if (event.action == KeyEvent.ACTION_DOWN && event.repeatCount == 0) {
-                val state = keyDispatcherState
-                state?.startTracking(event, this)
-                return true
-            } else if (event.action == KeyEvent.ACTION_UP) {
-                val state = keyDispatcherState
-                if (state != null && state.isTracking(event) && !event.isCanceled) {
-                    onSystemBack?.invoke()
-                    return true
-                }
-            }
-        }
-        return super.dispatchKeyEvent(event)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            return super.dispatchKeyEvent(event)
+        } else {
+            if (event.keyCode == KeyEvent.KEYCODE_BACK && onSystemBack != null) {
+                if (keyDispatcherState == null) {
+                    return super.dispatchKeyEvent(event)
+                }
+                if (event.action == KeyEvent.ACTION_DOWN && event.repeatCount == 0) {
+                    val state = keyDispatcherState
+                    state?.startTracking(event, this)
+                    return true
+                } else if (event.action == KeyEvent.ACTION_UP) {
+                    val state = keyDispatcherState
+                    if (state != null && state.isTracking(event) && !event.isCanceled) {
+                        onSystemBack?.invoke()
+                        return true
+                    }
+                }
+            }
+            return super.dispatchKeyEvent(event)
+        }
     }
 
-    fun updateParameters(
-        onSystemBack: (() -> Unit)?
-    ) {
+    fun updateParameters(onSystemBack: (() -> Unit)?) {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            findOnBackInvokedDispatcher()?.unregisterOnBackInvokedCallback(onBackInvokedCallback)
+        }
         this.onSystemBack = onSystemBack
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU && onSystemBack != null) {
+            findOnBackInvokedDispatcher()?.registerOnBackInvokedCallback(OnBackInvokedDispatcher.PRIORITY_OVERLAY, onBackInvokedCallback)
+        }
     }
 
     fun dismiss() {
         setViewTreeLifecycleOwner(null)
         decorView.removeView(this)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            findOnBackInvokedDispatcher()?.unregisterOnBackInvokedCallback(onBackInvokedCallback)
+        }
+    }
+
+    private val onBackInvokedCallback: OnBackInvokedCallback = OnBackInvokedCallback {
+        onSystemBack?.invoke()
     }
 }
```
Does make it so that if you got `android:enableOnBackInvokedCallback="true"` you properly dismiss the sheet instead of going back. This is on a device running API 35. Of course then turning it to `android:enableOnBackInvokedCallback="false"` again the opposite happens, we miss the callback since now the other path is taken, but since we check only for API level, and I do not think you can check for if `enableOnBackInvokedCallback` is true or false, we just miss it completely 😅

I did it this way since I was not sure if this library would like to bring in the dependency to `androidx.activity:activity`. I think in there one could then use this path https://developer.android.com/guide/navigation/custom-back/predictive-back-gesture#migrate-androidx instead of this path https://developer.android.com/guide/navigation/custom-back/predictive-back-gesture#migrate-app of the migration.

**Comment by StylianosGakis** [comment]
Now if we do try to add the activity dependency into the library, so we get access to the compat version of onBackInvoked, which is onBackPressed, and we add a patch like this:
```
Subject: [PATCH] Try with android.activity
---
Index: gradle/libs.versions.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/gradle/libs.versions.toml b/gradle/libs.versions.toml
--- a/gradle/libs.versions.toml	(revision 9df1393bfc0517554632f3d247f467e24f5de2dd)
+++ b/gradle/libs.versions.toml	(date 1719929795301)
@@ -6,7 +6,7 @@
 compose-bom = "2023.08.00"
 compose-compiler = "1.5.1"
 accompanist = "0.30.1"
-activity-compose = "1.7.2"
+activity-compose = "1.9.0"
 lifecycle = "2.6.1"
 
 plugin-android-gradle = "8.1.0"
Index: modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt b/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt
--- a/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt	(revision 9df1393bfc0517554632f3d247f467e24f5de2dd)
+++ b/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt	(date 1719930748579)
@@ -4,9 +4,10 @@
 import android.app.Activity
 import android.content.Context
 import android.content.ContextWrapper
-import android.view.KeyEvent
 import android.view.View
 import android.view.ViewGroup
+import androidx.activity.OnBackPressedCallback
+import androidx.activity.findViewTreeOnBackPressedDispatcherOwner
 import androidx.compose.foundation.layout.Box
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.CompositionContext
@@ -58,18 +59,18 @@
     val currentContent by rememberUpdatedState(content)
     val popupId = rememberSaveable { UUID.randomUUID() }
     val popupLayout = remember {
-        PopupLayout(
-            onSystemBack = onSystemBack,
-            composeView = view,
+            PopupLayout(
+                onSystemBack = onSystemBack,
+                composeView = view,
             popupId = popupId
-        ).apply {
-            setContent(parentComposition) {
-                Box(Modifier.semantics { this.popup() }) {
-                    currentContent()
-                }
-            }
-        }
-    }
+            ).apply {
+                setContent(parentComposition) {
+                    Box(Modifier.semantics { this.popup() }) {
+                        currentContent()
+                    }
+                }
+            }
+        }
 
     DisposableEffect(popupLayout) {
         popupLayout.show()
@@ -144,37 +145,31 @@
         content()
     }
 
-    @Suppress("ReturnCount")
-    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
-        if (event.keyCode == KeyEvent.KEYCODE_BACK && onSystemBack != null) {
-            if (keyDispatcherState == null) {
-                return super.dispatchKeyEvent(event)
-            }
-            if (event.action == KeyEvent.ACTION_DOWN && event.repeatCount == 0) {
-                val state = keyDispatcherState
-                state?.startTracking(event, this)
-                return true
-            } else if (event.action == KeyEvent.ACTION_UP) {
-                val state = keyDispatcherState
-                if (state != null && state.isTracking(event) && !event.isCanceled) {
-                    onSystemBack?.invoke()
-                    return true
-                }
-            }
-        }
-        return super.dispatchKeyEvent(event)
-    }
-
-    fun updateParameters(
-        onSystemBack: (() -> Unit)?
-    ) {
+    fun updateParameters(onSystemBack: (() -> Unit)?) {
+        onBackPressedCallback.remove()
         this.onSystemBack = onSystemBack
+        if (onSystemBack == null) return
+        val onBackPressedDispatcher = decorView.findViewTreeOnBackPressedDispatcherOwner()?.onBackPressedDispatcher
+        val lifecycleOwner = this.findViewTreeLifecycleOwner()
+        if (lifecycleOwner != null) {
+            onBackPressedDispatcher?.addCallback(lifecycleOwner, onBackPressedCallback)
+        } else {
+            onBackPressedDispatcher?.addCallback(onBackPressedCallback)
+        }
     }
 
     fun dismiss() {
         setViewTreeLifecycleOwner(null)
         decorView.removeView(this)
+        onBackPressedCallback.remove()
     }
+
+    private val onBackPressedCallback: OnBackPressedCallback =
+        object : OnBackPressedCallback(onSystemBack != null) {
+            override fun handleOnBackPressed() {
+                onSystemBack?.invoke()
+            }
+        }
 }
 
 private inline fun <reified T> findOwner(context: Context): T? {
Index: modalsheet/build.gradle.kts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/modalsheet/build.gradle.kts b/modalsheet/build.gradle.kts
--- a/modalsheet/build.gradle.kts	(revision 9df1393bfc0517554632f3d247f467e24f5de2dd)
+++ b/modalsheet/build.gradle.kts	(date 1719929818622)
@@ -35,6 +35,7 @@
     implementation(libs.compose.foundation)
     implementation(libs.compose.ui)
     implementation(libs.compose.material2)
+    implementation(libs.androidx.activitycompose)
     implementation(libs.androidx.savedstate)
     implementation(libs.androidx.viewmodel)
 }
```
This seems to work for both 
`android:enableOnBackInvokedCallback="true"`
and
`android:enableOnBackInvokedCallback="false"`
On API 35, since it always goes down the same path.

I also tried on API 30 as an example before any of this existed, and again it still does work for most samples.

---

It does not work on the "Custom Modal Sheet". But you got `cancellable = false` there. Then dismissing the sheet unfortunately seems to keep the back handler there still, since pressing back after dismissing it with the button does nothing. Probably something I missed.

It also does not work for "Sheet State Modal Sheet". Back here is forwarded to the outer BackHandler again. Not sure what is wrong here, perhaps something I am doing wrong with how I unregister and register it back when the `onSystemBack` callback is passed in again and again on `updateParameters`.

**Comment by StylianosGakis** [comment]
I think I might stop somewhere here and ask for your input regarding what you'd like to happen around this. Would you want the library to support the `enableOnBackInvokedCallback` cases? Do you want to add the activity dependency in here?
Do you want to perhaps even do it in a way where you could even support predictive back an have the dismissal happen in tandem with the back gesture progress?

---

One separate thought I had with all this, is that I think that normally in such Sheet cases, I wouldn't even want the `onSystemBack` parameter to be part of the API itself. I do not see a case where for example I would not want the sheet to intercept `back` in order to dismiss itself. And for scenarios where one wants the back button/gesture to do something special instead, a `(Predictive)BackHandler` inside the sheet content should be the way to go instead. And this does work perfectly well as the "Back Handler In Modal Sheet" demo item shows us.

---

So yes, looking forward to hearing back from you, until then I suppose users of this library would have to disable `enableOnBackInvokedCallback` for their entire app. So if anyone stumbles upon this, for now just do that.

**Comment by StylianosGakis** [comment]
https://github.com/StylianosGakis/modalsheet/pull/1
This is a very rough idea of how I imagine this working instead to avoid having to deal with any of this at all in the library and just defer to BackHandler which does the right thing for you.

The "custom modal sheet" sample still is a bit broken but again, I won't look into this further until I hear back from you. And the "Back Handler In Modal Sheet" example works perfectly fine since the BackHanlders guarantee that they fire in the order they are called.
I also didn't check what should happen if `cancellable` is set to false, but one could do many things, the simplest being that the BackHandler is always enabled when `cancellable` is `false`, and the lambda just does nothing. This is perfectly not-cancellable this way.

**Comment by StylianosGakis** [comment]
You are most definitely invited 👋

My only concern with that is that I've had so many bugs and problems with m3 that I have more or less lost hope that they will fix this properly, at least in the foreseeable future. I mean it is the reason I started looking into this library in the first place 😅

**Comment by wooodenleg** [comment]
This migration to dialog is fairly recent. Have you encountered bugs with that too? The evolution was Popup (like this lib) -> Window -> Dialog

**Comment by StylianosGakis** [comment]
I am using the latest stable version and I am experiencing the problems I mention. As far as I know the latest stable does not include this migration no.

Do you mean with this that perhaps after the latest changes in m3 drop, we should just use that and this library here won't really be needed anymore then?

**Comment by oleksandrbalan** [comment]
@StylianosGakis Wow, thanks for your time and investigation.

> Would you want the library to support the enableOnBackInvokedCallback cases?

Yes, I think it would be nice to have at least basic (without fancy animations) support for predictive back gesture. However, to be honest, I am a bit conservative and I am still happily using 3-buttons navigation on my device 😅 

 > Do you want to add the activity dependency in here?
 
 Yeah, I think it is a valid use-case to add it to the lib dependencies 👍 
 
 > I wouldn't even want the onSystemBack parameter to be part of the API itself.
 
 There was a use-case on one recent project where we have been using complex navigation inside bottom sheet (God bless designers), and there we needed to disable system back and override it with custom behavior. So I would like to keep it there, see #25.
 
 > The "custom modal sheet" sample still is a bit broken.
 
 Oopsie, looks like it was broken since last release 🤦  
 
 I will take a look on it.
 
 > https://github.com/StylianosGakis/modalsheet/pull/1 This is a very rough idea of how I imagine this working.
 
Thanks, I am now also playing around with an idea to slam `BackHandler` right into `PopupLayout` content and remove everything related to the `onSystemBack` from it. During my monkey-testing it looks fine, but I will be really grateful if you can check if it works for you. I will open a PR later this evening.
 
 <details>
  <summary>Patch for FullscreenPopup</summary>
  
  ```kotlin
Index: modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt b/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt
--- a/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt	(revision 55d0c458f612dc2fb1c601d7b45333aee2c97900)
+++ b/modalsheet/src/main/java/eu/wewox/modalsheet/FullscreenPopup.kt	(date 1720032656720)
@@ -4,14 +4,13 @@
 import android.app.Activity
 import android.content.Context
 import android.content.ContextWrapper
-import android.view.KeyEvent
 import android.view.View
 import android.view.ViewGroup
+import androidx.activity.compose.BackHandler
 import androidx.compose.foundation.layout.Box
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.CompositionContext
 import androidx.compose.runtime.DisposableEffect
-import androidx.compose.runtime.SideEffect
 import androidx.compose.runtime.getValue
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.remember
@@ -59,11 +58,13 @@
     val popupId = rememberSaveable { UUID.randomUUID() }
     val popupLayout = remember {
         PopupLayout(
-            onSystemBack = onSystemBack,
             composeView = view,
-            popupId = popupId
+            popupId = popupId,
         ).apply {
             setContent(parentComposition) {
+                BackHandler(onSystemBack != null) {
+                    onSystemBack?.invoke()
+                }
                 Box(Modifier.semantics { this.popup() }) {
                     currentContent()
                 }
@@ -73,21 +74,12 @@
 
     DisposableEffect(popupLayout) {
         popupLayout.show()
-        popupLayout.updateParameters(
-            onSystemBack = onSystemBack
-        )
         onDispose {
             popupLayout.disposeComposition()
             // Remove the window
             popupLayout.dismiss()
         }
     }
-
-    SideEffect {
-        popupLayout.updateParameters(
-            onSystemBack = onSystemBack
-        )
-    }
 }
 
 /**
@@ -95,7 +87,6 @@
  */
 @SuppressLint("ViewConstructor")
 private class PopupLayout(
-    private var onSystemBack: (() -> Unit)?,
     composeView: View,
     popupId: UUID
 ) : AbstractComposeView(composeView.context),
@@ -144,33 +135,6 @@
         content()
     }
 
-    @Suppress("ReturnCount")
-    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
-        if (event.keyCode == KeyEvent.KEYCODE_BACK && onSystemBack != null) {
-            if (keyDispatcherState == null) {
-                return super.dispatchKeyEvent(event)
-            }
-            if (event.action == KeyEvent.ACTION_DOWN && event.repeatCount == 0) {
-                val state = keyDispatcherState
-                state?.startTracking(event, this)
-                return true
-            } else if (event.action == KeyEvent.ACTION_UP) {
-                val state = keyDispatcherState
-                if (state != null && state.isTracking(event) && !event.isCanceled) {
-                    onSystemBack?.invoke()
-                    return true
-                }
-            }
-        }
-        return super.dispatchKeyEvent(event)
-    }
-
-    fun updateParameters(
-        onSystemBack: (() -> Unit)?
-    ) {
-        this.onSystemBack = onSystemBack
-    }
-
     fun dismiss() {
         setViewTreeLifecycleOwner(null)
         decorView.removeView(this)
```
</details>
 
 
@wooodenleg Huh, it is very interesting that they have changed so many underlying implementations 🤔 Maybe once Material team will come up with fullscreen popup 😏 Jokes aside, I would like to KISS it as for now and stick with something which works for us 🤷‍♂️

**Comment by StylianosGakis** [comment]
Just got a chance to look at this again. Thanks a lot for reading all of my rambling here, and for taking the initiative to fix this!
Yeah when we can rely on their re-implementation of it using a Dialog that'd be great, but for now if we can fix it right here for our current needs that'd be even better!

Love that you did what I did only so much simpler in that PR 😅 It looks good from looking at it, and if all the samples work well then that sounds perfect. I didn't try this out in my project to see it working with whatever else I may be doing there, do you have a way to publish perhaps beta versions or something like that so I can test this out in the real project too?

I could try to copy-paste the entire implementation and try to do it that way instead if you can't do that, but that'd take me a bit more time before I can jump into it :D

**Comment by oleksandrbalan** [comment]
Sorry for the late reply, I have managed to publish [0.7.0-alpha01](https://github.com/oleksandrbalan/modalsheet/releases/tag/v0.7.0-alpha01) version yesterday with a changes from that PR. 

Please let me know if it works for you 👀

**Comment by StylianosGakis** [comment]
Thanks a lot! I will check this out a bit later this week, a bit cramped with some other stuff right now. I really appreciate the quick release for this 🙏

**Comment by oleksandrbalan** [comment]
Awesome, thx!

I have released [1.7.0-alpha02](https://github.com/oleksandrbalan/modalsheet/releases/tag/v0.7.0-alpha02) with fixes regarding non-cancelable sheet which Filip mentioned in review, so better grab that version 😄

**Comment by StylianosGakis** [comment]
Had a little vacation in-between, sorry about that. Checked right now and alpha release 02 works perfectly fine for me with enableOnBackInvokedCallback on Android 15.
Super nice, thank you so much for addressing this!

**Comment by oleksandrbalan** [comment]
No problem 😄 
Thx, for checking 👍 

I have released final [0.7.0](https://github.com/oleksandrbalan/modalsheet/releases/tag/v0.7.0) 🚀
